## Apriori Algorithm

Question: Can we predict which products a customer will most likely purchase together within various product segments?

Goal: Help Amazon identify products frequently bought together by customers to increase sales and revenues (cross sell) by analyzing Amazon Marketplace segment data.

In [1]:
# Import dependencies
from sqlalchemy import create_engine
import pandas as pd

# Importing apriori dependencies
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [2]:
# Connect to postgres/RDS database 
POSTGRES_ADDRESS = 'mypostgresdb.cwuhtytzosg8.us-east-2.rds.amazonaws.com' 
POSTGRES_PORT = '5432'
POSTGRES_USERNAME = 'root' 
POSTGRES_PASSWORD = 'XXXX' 
POSTGRES_DBNAME = 'my_data_class_db'

# A long string that contains the necessary Postgres login information
postgres_str = ('postgresql://{username}:{password}@{ipaddress}:{port}/{dbname}'
.format(username=POSTGRES_USERNAME,
password=POSTGRES_PASSWORD,
ipaddress=POSTGRES_ADDRESS,
port=POSTGRES_PORT,
dbname=POSTGRES_DBNAME))
# Create the connection
cnx = create_engine(postgres_str)

### Video Analysis 

In [3]:
# Load database for sentiment/topic analysis
df_videos = pd.read_sql_query('''SELECT * FROM videos_apriori_analysis''', con=cnx)
df_videos.head()

customer_id  product_id  quantity
0     25551507  0788812807         1
1     25551507  6302320402         1
2     31354506  6301442733         1
3     42622115  B00003CX7L         1
4     27446106  0788806270         1

In [5]:
# Create pivot table to run algorithm 
apriori_table = df_videos.groupby(["customer_id","product_id"])["quantity"].sum().unstack().reset_index().fillna(0).set_index("customer_id")
apriori_table

product_id   0738920525  0767800958  0767815963  076781598X  076783822X  \
customer_id                                                               
15160               0.0         0.0         0.0         0.0         0.0   
19893               0.0         0.0         0.0         0.0         0.0   
20767               0.0         0.0         0.0         0.0         0.0   
26626               0.0         0.0         0.0         0.0         0.0   
41909               0.0         0.0         0.0         0.0         0.0   
...                 ...         ...         ...         ...         ...   
53094662            0.0         0.0         0.0         0.0         0.0   
53094728            0.0         0.0         0.0         0.0         0.0   
53095639            0.0         0.0         0.0         0.0         0.0   
53096090            0.0         0.0         0.0         0.0         0.0   
53096171            0.0         0.0         0.0         0.0         0.0   

product_id   0773386777  0780614097  078062128X  0780625900  0780626028  ...  \
customer_id                                                              ...   
15160               0.0         0.0         0.0         0.0         0.0  ...   
19893               0.0         0.0         0.0         0.0         0.0  ...   
20767               0.0         0.0         0.0         0.0         0.0  ...   
26626               0.0         0.0         0.0         0.0         0.0  ...   
41909               0.0         0.0         0.0         0.0         0.0  ...   
...                 ...         ...         ...         ...         ...  ...   
53094662            0.0         0.0         0.0         0.0         0.0  ...   
53094728            0.0         0.0         0.0         0.0         0.0  ...   
53095639            0.0         0.0         0.0         0.0         0.0  ...   
53096090            0.0         0.0         0.0         0.0         0.0  ...   
53096171            0.0         0.0         0.0         0.0         0.0  ...   

product_id   B000E4C2TY  B000FP5PPA  B000H61X62  B000QX1SUC  B000TOSN7Y  \
customer_id                                                               
15160               0.0         0.0         0.0         0.0         0.0   
19893               0.0         0.0         0.0         0.0         0.0   
20767               0.0         0.0         0.0         0.0         0.0   
26626               0.0         0.0         0.0         0.0         0.0   
41909               0.0         0.0         0.0         0.0         0.0   
...                 ...         ...         ...         ...         ...   
53094662            0.0         0.0         0.0         0.0         0.0   
53094728            0.0         0.0         0.0         0.0         0.0   
53095639            0.0         0.0         0.0         0.0         0.0   
53096090            0.0         0.0         0.0         0.0         0.0   
53096171            0.0         0.0         0.0         0.0         0.0   

product_id   B0019KBJOQ  B004J0SG6C  B00AVPYKFU  B00B9LNBWS  B00BBND56G  
customer_id                                                              
15160               0.0         0.0         0.0         0.0         0.0  
19893               0.0         0.0         0.0         0.0         0.0  
20767               0.0         0.0         0.0         0.0         0.0  
26626               0.0         0.0         0.0         0.0         0.0  
41909               0.0         0.0         0.0         0.0         0.0  
...                 ...         ...         ...         ...         ...  
53094662            0.0         0.0         0.0         0.0         0.0  
53094728            0.0         0.0         0.0         0.0         0.0  
53095639            0.0         0.0         0.0         0.0         0.0  
53096090            0.0         0.0         0.0         0.0         0.0  
53096171            0.0         0.0         0.0         0.0         0.0  

[12104 rows x 750 columns]

In [6]:
# Cleaning Apriori table to remove unnecessary numbers
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1
apriori_cleaned_videos = apriori_table.applymap(encode_units)
apriori_cleaned_videos

product_id   0738920525  0767800958  0767815963  076781598X  076783822X  \
customer_id                                                               
15160                 0           0           0           0           0   
19893                 0           0           0           0           0   
20767                 0           0           0           0           0   
26626                 0           0           0           0           0   
41909                 0           0           0           0           0   
...                 ...         ...         ...         ...         ...   
53094662              0           0           0           0           0   
53094728              0           0           0           0           0   
53095639              0           0           0           0           0   
53096090              0           0           0           0           0   
53096171              0           0           0           0           0   

product_id   0773386777  0780614097  078062128X  0780625900  0780626028  ...  \
customer_id                                                              ...   
15160                 0           0           0           0           0  ...   
19893                 0           0           0           0           0  ...   
20767                 0           0           0           0           0  ...   
26626                 0           0           0           0           0  ...   
41909                 0           0           0           0           0  ...   
...                 ...         ...         ...         ...         ...  ...   
53094662              0           0           0           0           0  ...   
53094728              0           0           0           0           0  ...   
53095639              0           0           0           0           0  ...   
53096090              0           0           0           0           0  ...   
53096171              0           0           0           0           0  ...   

product_id   B000E4C2TY  B000FP5PPA  B000H61X62  B000QX1SUC  B000TOSN7Y  \
customer_id                                                               
15160                 0           0           0           0           0   
19893                 0           0           0           0           0   
20767                 0           0           0           0           0   
26626                 0           0           0           0           0   
41909                 0           0           0           0           0   
...                 ...         ...         ...         ...         ...   
53094662              0           0           0           0           0   
53094728              0           0           0           0           0   
53095639              0           0           0           0           0   
53096090              0           0           0           0           0   
53096171              0           0           0           0           0   

product_id   B0019KBJOQ  B004J0SG6C  B00AVPYKFU  B00B9LNBWS  B00BBND56G  
customer_id                                                              
15160                 0           0           0           0           0  
19893                 0           0           0           0           0  
20767                 0           0           0           0           0  
26626                 0           0           0           0           0  
41909                 0           0           0           0           0  
...                 ...         ...         ...         ...         ...  
53094662              0           0           0           0           0  
53094728              0           0           0           0           0  
53095639              0           0           0           0           0  
53096090              0           0           0           0           0  
53096171              0           0           0           0           0  

[12104 rows x 750 columns]

In [7]:
# Running apriori algorithm on cleaned dataset. 
item_association_videos = apriori(apriori_cleaned_videos, min_support=0.0001, use_colnames=True)
item_association_videos

support                                           itemsets
0     0.002396                                       (0738920525)
1     0.001322                                       (0767800958)
2     0.001652                                       (0767815963)
3     0.001652                                       (076781598X)
4     0.002479                                       (076783822X)
...        ...                                                ...
1615  0.000165  (0788812408, 0788805533, 155890641X, 630027419...
1616  0.000165  (0788812408, 0788805533, 155890641X, 630027419...
1617  0.000165  (0788812408, 155890641X, 6300274195, 078880219...
1618  0.000165  (0788812408, 0788805533, 155890641X, 630027419...
1619  0.000165  (0788812408, 0788805533, 155890641X, 630027419...

[1620 rows x 2 columns]

In [8]:
# Apriori association results table and confidence levels. 
apriori_rules_videos = association_rules(item_association_videos, metric="lift", min_threshold=1)

In [9]:
# Show output by descending order of confidence level
apriori_rules_videos = apriori_rules_videos.sort_values(["confidence"], ascending=False)
apriori_rules_videos 

antecedents  \
9175                (0788802194, 0788812408, 0788812807)   
11817  (0788812408, 155890641X, 0788802194, 630440113...   
11819  (155890641X, 6300274195, 0788802194, 630440113...   
11820   (155890641X, 0788812408, 0788802194, 6300274195)   
11821   (155890641X, 0788812408, 6304401132, 6300274195)   
...                                                  ...   
466                                         (B00004U8H5)   
322                                         (6302208661)   
430                                         (B00004U8H5)   
174                                         (094567189X)   
172                                         (094567189X)   

                                consequents  antecedent support  \
9175   (6304401132, 0788805533, 0788806270)            0.000165   
11817                          (6300274195)            0.000165   
11819                          (0788812408)            0.000165   
11820              (6304401132, 0788812807)            0.000165   
11821              (0788802194, 0788812807)            0.000165   
...                                     ...                 ...   
466                            (B000083C59)            0.004957   
322                            (6302967945)            0.004957   
430                            (6305107807)            0.004957   
174                            (6303243606)            0.005040   
172                            (0945671636)            0.005040   

       consequent support   support  confidence         lift  leverage  \
9175             0.000165  0.000165    1.000000  6052.000000  0.000165   
11817            0.001157  0.000165    1.000000   864.571429  0.000165   
11819            0.003139  0.000165    1.000000   318.526316  0.000165   
11820            0.000165  0.000165    1.000000  6052.000000  0.000165   
11821            0.000165  0.000165    1.000000  6052.000000  0.000165   
...                   ...       ...         ...          ...       ...   
466              0.002809  0.000165    0.033333    11.866667  0.000151   
322              0.002726  0.000165    0.033333    12.226263  0.000152   
430              0.001404  0.000165    0.033333    23.733333  0.000158   
174              0.002892  0.000165    0.032787    11.338642  0.000151   
172              0.002313  0.000165    0.032787    14.173302  0.000154   

       conviction  
9175          inf  
11817         inf  
11819         inf  
11820         inf  
11821         inf  
...           ...  
466      1.031577  
322      1.031662  
430      1.033030  
174      1.030909  
172      1.031507  

[21504 rows x 9 columns]

Interpreataion of 1st row of results, VHS bought together:

0788802194: Cinderella

0788812408: The Little Mermaid

0788812807: Lady and the Tramp

6304401132: Sleeping Beauty

0788805533: The Lion King II

0788806270: Bambi

In [12]:
# Add category column (to be used in viz)
apriori_rules_videos['Category'] = 'Videos'
apriori_rules_videos

antecedents  \
9175                (0788802194, 0788812408, 0788812807)   
11817  (0788812408, 155890641X, 0788802194, 630440113...   
11819  (155890641X, 6300274195, 0788802194, 630440113...   
11820   (155890641X, 0788812408, 0788802194, 6300274195)   
11821   (155890641X, 0788812408, 6304401132, 6300274195)   
...                                                  ...   
466                                         (B00004U8H5)   
322                                         (6302208661)   
430                                         (B00004U8H5)   
174                                         (094567189X)   
172                                         (094567189X)   

                                consequents  antecedent support  \
9175   (6304401132, 0788805533, 0788806270)            0.000165   
11817                          (6300274195)            0.000165   
11819                          (0788812408)            0.000165   
11820              (6304401132, 0788812807)            0.000165   
11821              (0788802194, 0788812807)            0.000165   
...                                     ...                 ...   
466                            (B000083C59)            0.004957   
322                            (6302967945)            0.004957   
430                            (6305107807)            0.004957   
174                            (6303243606)            0.005040   
172                            (0945671636)            0.005040   

       consequent support   support  confidence         lift  leverage  \
9175             0.000165  0.000165    1.000000  6052.000000  0.000165   
11817            0.001157  0.000165    1.000000   864.571429  0.000165   
11819            0.003139  0.000165    1.000000   318.526316  0.000165   
11820            0.000165  0.000165    1.000000  6052.000000  0.000165   
11821            0.000165  0.000165    1.000000  6052.000000  0.000165   
...                   ...       ...         ...          ...       ...   
466              0.002809  0.000165    0.033333    11.866667  0.000151   
322              0.002726  0.000165    0.033333    12.226263  0.000152   
430              0.001404  0.000165    0.033333    23.733333  0.000158   
174              0.002892  0.000165    0.032787    11.338642  0.000151   
172              0.002313  0.000165    0.032787    14.173302  0.000154   

       conviction Category  
9175          inf   Videos  
11817         inf   Videos  
11819         inf   Videos  
11820         inf   Videos  
11821         inf   Videos  
...           ...      ...  
466      1.031577   Videos  
322      1.031662   Videos  
430      1.033030   Videos  
174      1.030909   Videos  
172      1.031507   Videos  

[21504 rows x 10 columns]

In [13]:
apriori_rules_videos.to_csv(r'C:\Users\li_mi\Class\final_project\dev\apriori_rules_videos.csv', index = False)

### Personal Care Analysis

In [14]:
# Get data
df_personal_care = pd.read_sql_query('''SELECT * FROM personal_care_appliances_apriori_analysis''', con=cnx)
df_personal_care.head()

customer_id  product_id  quantity
0     32114233  B00OYRW4UE         1
1     18125776  B0000537JQ         1
2     19917519  B00HXXO332         1
3     18277171  B00EOB0JA2         1
4      2592955  B00HES9CMS         1

In [15]:
# Create pivot table to run algorithm 
apriori_table_personal_care = df_personal_care.groupby(["customer_id","product_id"])["quantity"].sum().unstack().reset_index().fillna(0).set_index("customer_id")
apriori_table_personal_care

product_id   B0000532OT  B0000532OV  B0000537JP  B0000537JQ  B000068PBJ  \
customer_id                                                               
10470               0.0         0.0         0.0         0.0         0.0   
11344               0.0         0.0         0.0         0.0         0.0   
12674               0.0         0.0         0.0         0.0         0.0   
13044               0.0         0.0         0.0         0.0         0.0   
14147               0.0         0.0         0.0         0.0         0.0   
...                 ...         ...         ...         ...         ...   
53092127            0.0         0.0         0.0         0.0         0.0   
53094082            0.0         0.0         0.0         0.0         0.0   
53094709            0.0         0.0         0.0         0.0         0.0   
53095826            0.0         0.0         0.0         0.0         0.0   
53096344            0.0         0.0         0.0         0.0         0.0   

product_id   B00008J1ZZ  B00009RB1I  B0000YS1BG  B0000YXUW6  B00012FJ9O  ...  \
customer_id                                                              ...   
10470               0.0         0.0         0.0         0.0         0.0  ...   
11344               0.0         0.0         0.0         0.0         0.0  ...   
12674               0.0         0.0         0.0         0.0         0.0  ...   
13044               0.0         0.0         0.0         0.0         0.0  ...   
14147               0.0         0.0         0.0         0.0         0.0  ...   
...                 ...         ...         ...         ...         ...  ...   
53092127            0.0         0.0         0.0         0.0         0.0  ...   
53094082            0.0         0.0         0.0         0.0         0.0  ...   
53094709            0.0         0.0         0.0         0.0         0.0  ...   
53095826            0.0         0.0         0.0         0.0         0.0  ...   
53096344            0.0         0.0         0.0         0.0         0.0  ...   

product_id   B00P6TUO5G  B00QH8QJ2C  B00QH96JQC  B00QH9M6QY  B00QR4JRHU  \
customer_id                                                               
10470               0.0         0.0         0.0         0.0         0.0   
11344               0.0         0.0         0.0         0.0         0.0   
12674               0.0         0.0         0.0         0.0         0.0   
13044               0.0         0.0         0.0         0.0         0.0   
14147               0.0         0.0         0.0         0.0         0.0   
...                 ...         ...         ...         ...         ...   
53092127            0.0         0.0         0.0         0.0         0.0   
53094082            0.0         0.0         0.0         0.0         0.0   
53094709            0.0         0.0         0.0         0.0         0.0   
53095826            0.0         0.0         0.0         0.0         0.0   
53096344            0.0         0.0         0.0         0.0         0.0   

product_id   B00R3PFF4Q  B00RWIWFFQ  B00S02EJZW  B00TOYNBA4  B00XZJ2G46  
customer_id                                                              
10470               0.0         0.0         0.0         0.0         0.0  
11344               0.0         0.0         0.0         0.0         0.0  
12674               0.0         0.0         0.0         0.0         0.0  
13044               0.0         0.0         0.0         0.0         0.0  
14147               0.0         0.0         0.0         0.0         0.0  
...                 ...         ...         ...         ...         ...  
53092127            0.0         0.0         0.0         0.0         0.0  
53094082            0.0         0.0         0.0         0.0         0.0  
53094709            0.0         0.0         0.0         0.0         0.0  
53095826            0.0         0.0         0.0         0.0         0.0  
53096344            0.0         0.0         0.0         0.0         0.0  

[33555 rows x 400 columns]

In [16]:
# Cleaning Apriori table to remove unnecessary numbers
apriori_cleaned_pcare = apriori_table_personal_care.applymap(encode_units)

In [17]:
# Running apriori algorithm on cleaned dataset. 
item_association_pcare = apriori(apriori_cleaned_pcare, min_support=0.0001, use_colnames=True)
item_association_pcare

support                  itemsets
0    0.000834              (B0000532OT)
1    0.001252              (B0000532OV)
2    0.001371              (B0000537JP)
3    0.005275              (B0000537JQ)
4    0.002176              (B000068PBJ)
..        ...                       ...
405  0.000119  (B004O23YE4, B00H8ZTLLY)
406  0.000149  (B004O25RM6, B004O25RJ4)
407  0.000119  (B00H8ZTLLY, B004O25RJ4)
408  0.000119  (B004O275X0, B004O2762U)
409  0.000119  (B00H8ZTLLY, B004O276PW)

[410 rows x 2 columns]

In [22]:
# Apriori association results table and confidence levels. 
apriori_rules_pcare = association_rules(item_association_pcare, metric="lift", min_threshold=1)

In [23]:
# Show output by descending order of confidence level
apriori_rules_pcare = apriori_rules_pcare.sort_values(["confidence"], ascending=False)
apriori_rules_pcare 

antecedents   consequents  antecedent support  consequent support  \
12  (B004O25RM6)  (B004O25RJ4)            0.000924            0.001699   
8   (B004O23YE4)  (B004O276PW)            0.004053            0.009715   
13  (B004O25RJ4)  (B004O25RM6)            0.001699            0.000924   
17  (B004O2762U)  (B004O275X0)            0.001609            0.001997   
0   (B000FFYC6S)  (B000Y9PERQ)            0.002474            0.005037   
15  (B004O25RJ4)  (B00H8ZTLLY)            0.001699            0.003993   
16  (B004O275X0)  (B004O2762U)            0.001997            0.001609   
3   (B004O23Y9E)  (B004O276PW)            0.003576            0.009715   
4   (B004PV7PIK)  (B004O23Y9E)            0.002891            0.003576   
9   (B004O276PW)  (B004O23YE4)            0.009715            0.004053   
1   (B000Y9PERQ)  (B000FFYC6S)            0.005037            0.002474   
5   (B004O23Y9E)  (B004PV7PIK)            0.003576            0.002891   
7   (B004O23Y9E)  (B00H8ZTLLY)            0.003576            0.003993   
6   (B00H8ZTLLY)  (B004O23Y9E)            0.003993            0.003576   
11  (B00H8ZTLLY)  (B004O23YE4)            0.003993            0.004053   
18  (B00H8ZTLLY)  (B004O276PW)            0.003993            0.009715   
14  (B00H8ZTLLY)  (B004O25RJ4)            0.003993            0.001699   
10  (B004O23YE4)  (B00H8ZTLLY)            0.004053            0.003993   
2   (B004O276PW)  (B004O23Y9E)            0.009715            0.003576   
19  (B004O276PW)  (B00H8ZTLLY)            0.009715            0.003993   

     support  confidence       lift  leverage  conviction  
12  0.000149    0.161290  94.949066  0.000147    1.190282  
8   0.000358    0.088235   9.082010  0.000318    1.086119  
13  0.000149    0.087719  94.949066  0.000147    1.095141  
17  0.000119    0.074074  37.097844  0.000116    1.077844  
0   0.000179    0.072289  14.353033  0.000166    1.072493  
15  0.000119    0.070175  17.572663  0.000112    1.071177  
16  0.000119    0.059701  37.097844  0.000116    1.061781  
3   0.000149    0.041667   4.288727  0.000114    1.033340  
4   0.000119    0.041237  11.530928  0.000109    1.039281  
9   0.000358    0.036810   9.082010  0.000318    1.034009  
1   0.000179    0.035503  14.353033  0.000166    1.034245  
5   0.000119    0.033333  11.530928  0.000109    1.031492  
7   0.000119    0.033333   8.347015  0.000105    1.030352  
6   0.000119    0.029851   8.347015  0.000105    1.027083  
11  0.000119    0.029851   7.365013  0.000103    1.026591  
18  0.000119    0.029851   3.072521  0.000080    1.020755  
14  0.000119    0.029851  17.572663  0.000112    1.029018  
10  0.000119    0.029412   7.365013  0.000103    1.026189  
2   0.000149    0.015337   4.288727  0.000114    1.011944  
19  0.000119    0.012270   3.072521  0.000080    1.008379

In [24]:
# Add category column (to be used in viz)
apriori_rules_pcare['Category'] = 'Personal Care'
apriori_rules_pcare

antecedents   consequents  antecedent support  consequent support  \
12  (B004O25RM6)  (B004O25RJ4)            0.000924            0.001699   
8   (B004O23YE4)  (B004O276PW)            0.004053            0.009715   
13  (B004O25RJ4)  (B004O25RM6)            0.001699            0.000924   
17  (B004O2762U)  (B004O275X0)            0.001609            0.001997   
0   (B000FFYC6S)  (B000Y9PERQ)            0.002474            0.005037   
15  (B004O25RJ4)  (B00H8ZTLLY)            0.001699            0.003993   
16  (B004O275X0)  (B004O2762U)            0.001997            0.001609   
3   (B004O23Y9E)  (B004O276PW)            0.003576            0.009715   
4   (B004PV7PIK)  (B004O23Y9E)            0.002891            0.003576   
9   (B004O276PW)  (B004O23YE4)            0.009715            0.004053   
1   (B000Y9PERQ)  (B000FFYC6S)            0.005037            0.002474   
5   (B004O23Y9E)  (B004PV7PIK)            0.003576            0.002891   
7   (B004O23Y9E)  (B00H8ZTLLY)            0.003576            0.003993   
6   (B00H8ZTLLY)  (B004O23Y9E)            0.003993            0.003576   
11  (B00H8ZTLLY)  (B004O23YE4)            0.003993            0.004053   
18  (B00H8ZTLLY)  (B004O276PW)            0.003993            0.009715   
14  (B00H8ZTLLY)  (B004O25RJ4)            0.003993            0.001699   
10  (B004O23YE4)  (B00H8ZTLLY)            0.004053            0.003993   
2   (B004O276PW)  (B004O23Y9E)            0.009715            0.003576   
19  (B004O276PW)  (B00H8ZTLLY)            0.009715            0.003993   

     support  confidence       lift  leverage  conviction       Category  
12  0.000149    0.161290  94.949066  0.000147    1.190282  Personal Care  
8   0.000358    0.088235   9.082010  0.000318    1.086119  Personal Care  
13  0.000149    0.087719  94.949066  0.000147    1.095141  Personal Care  
17  0.000119    0.074074  37.097844  0.000116    1.077844  Personal Care  
0   0.000179    0.072289  14.353033  0.000166    1.072493  Personal Care  
15  0.000119    0.070175  17.572663  0.000112    1.071177  Personal Care  
16  0.000119    0.059701  37.097844  0.000116    1.061781  Personal Care  
3   0.000149    0.041667   4.288727  0.000114    1.033340  Personal Care  
4   0.000119    0.041237  11.530928  0.000109    1.039281  Personal Care  
9   0.000358    0.036810   9.082010  0.000318    1.034009  Personal Care  
1   0.000179    0.035503  14.353033  0.000166    1.034245  Personal Care  
5   0.000119    0.033333  11.530928  0.000109    1.031492  Personal Care  
7   0.000119    0.033333   8.347015  0.000105    1.030352  Personal Care  
6   0.000119    0.029851   8.347015  0.000105    1.027083  Personal Care  
11  0.000119    0.029851   7.365013  0.000103    1.026591  Personal Care  
18  0.000119    0.029851   3.072521  0.000080    1.020755  Personal Care  
14  0.000119    0.029851  17.572663  0.000112    1.029018  Personal Care  
10  0.000119    0.029412   7.365013  0.000103    1.026189  Personal Care  
2   0.000149    0.015337   4.288727  0.000114    1.011944  Personal Care  
19  0.000119    0.012270   3.072521  0.000080    1.008379  Personal Care

In [25]:
apriori_rules_pcare.to_csv(r'C:\Users\li_mi\Class\final_project\dev\apriori_rules_personal_care.csv', index = False)

Interpretaion of 1st row of results:

B004O25RM6: doTERRA - White Fir Essential Oil - Helps to Balance Emotions and Soothe Anxious Feelings, Provides Soothing Effect in Massage, Relaxing Aroma; For Diffusion, Internal, or Topical Use - 15 ml

B004O25RJ4: doTERRA - Rosemary Essential Oil - 15 mL


### Video games Analysis 

In [26]:
#  Upload df
df_video_games = pd.read_sql_query('''SELECT * FROM video_games_apriori_analysis''', con=cnx)
df_video_games.head()

customer_id  product_id  quantity
0     12039526  B001CXYMFS         1
1     48880662  B0053OLY9O         1
2     45205407  B00KVP78FE         1
3     10548951  B00273Z9WM         1
4     50165446  B003O6E800         1

In [27]:
# Create pivot table to run algorithm 
apriori_video_games = df_video_games.groupby(["customer_id","product_id"])["quantity"].sum().unstack().reset_index().fillna(0).set_index("customer_id")
apriori_video_games

product_id   B0009VXBAQ  B000ERVMI8  B000FQ9R4E  B000M4KIME  B000NUBY0C  \
customer_id                                                               
10018               0.0         0.0         0.0         0.0         0.0   
10481               0.0         0.0         0.0         0.0         0.0   
10670               0.0         0.0         0.0         0.0         0.0   
10866               0.0         0.0         0.0         0.0         0.0   
11026               0.0         0.0         0.0         0.0         0.0   
...                 ...         ...         ...         ...         ...   
53092633            0.0         0.0         0.0         0.0         0.0   
53092767            0.0         0.0         0.0         0.0         0.0   
53093124            0.0         0.0         0.0         0.0         0.0   
53093730            0.0         0.0         0.0         0.0         0.0   
53094440            1.0         0.0         0.0         0.0         0.0   

product_id   B000OYMYZQ  B000XJNTNS  B000ZKA0J6  B0013OL0BK  B0015AARJI  ...  \
customer_id                                                              ...   
10018               0.0         0.0         0.0         0.0         0.0  ...   
10481               0.0         0.0         0.0         0.0         0.0  ...   
10670               0.0         0.0         0.0         0.0         0.0  ...   
10866               0.0         0.0         0.0         0.0         0.0  ...   
11026               0.0         0.0         0.0         0.0         0.0  ...   
...                 ...         ...         ...         ...         ...  ...   
53092633            0.0         1.0         0.0         0.0         0.0  ...   
53092767            0.0         0.0         0.0         0.0         0.0  ...   
53093124            0.0         0.0         0.0         0.0         0.0  ...   
53093730            0.0         0.0         0.0         0.0         0.0  ...   
53094440            0.0         0.0         0.0         0.0         0.0  ...   

product_id   B00JK00S0S  B00KSQHX1K  B00KTNSKZU  B00KVOVBGM  B00KVP78FE  \
customer_id                                                               
10018               0.0         0.0         0.0         0.0         0.0   
10481               0.0         0.0         0.0         0.0         0.0   
10670               0.0         0.0         0.0         0.0         0.0   
10866               0.0         0.0         0.0         0.0         0.0   
11026               0.0         0.0         0.0         0.0         0.0   
...                 ...         ...         ...         ...         ...   
53092633            0.0         0.0         0.0         0.0         0.0   
53092767            0.0         0.0         0.0         0.0         0.0   
53093124            0.0         0.0         0.0         0.0         0.0   
53093730            0.0         0.0         0.0         0.0         0.0   
53094440            0.0         0.0         0.0         0.0         0.0   

product_id   B00KVSQ848  B00MU1YENG  B00NFXON1Q  B00O9JLAX4  B00RSXRLUE  
customer_id                                                              
10018               0.0         0.0         0.0         0.0         0.0  
10481               0.0         0.0         0.0         0.0         0.0  
10670               0.0         0.0         0.0         0.0         0.0  
10866               1.0         0.0         0.0         0.0         0.0  
11026               0.0         0.0         0.0         0.0         0.0  
...                 ...         ...         ...         ...         ...  
53092633            0.0         0.0         0.0         0.0         0.0  
53092767            0.0         0.0         0.0         0.0         0.0  
53093124            0.0         0.0         0.0         0.0         0.0  
53093730            0.0         0.0         0.0         0.0         0.0  
53094440            0.0         0.0         0.0         0.0         0.0  

[136234 rows x 125 columns]

In [28]:
# Cleaning Apriori table to remove unnecessary numbers
apriori_video_games = apriori_video_games.applymap(encode_units)

In [29]:
# Running apriori algorithm on cleaned dataset. 
item_association_video_games = apriori(apriori_video_games, min_support=0.0001, use_colnames=True)

In [30]:
#  Apriori association results table and confidence levels. 
apriori_rules_video_games = association_rules(item_association_video_games, metric="lift", min_threshold=1)

In [31]:
# Show output by descending order of confidence level
apriori_rules_video_games  = apriori_rules_video_games.sort_values(["confidence"], ascending=False)
apriori_rules_video_games

antecedents               consequents  antecedent support  \
692  (B00BGA9X9W, B00JK00S0S)              (B00BGA9WK2)            0.000213   
674  (B00BGA9X9W, B00BI83EVU)              (B00BGA9WK2)            0.000308   
668  (B00BGA9X9W, B00BGAA3S2)              (B00BGA9WK2)            0.000896   
661  (B003O6FV8S, B0086V5UF0)              (B0053BG26C)            0.000257   
679  (B00BGA9WK2, B00ENFVJJO)              (B00BGA9X9W)            0.000749   
..                        ...                       ...                 ...   
657              (B0050SXKU4)  (B003O6CBIG, B007XVTR5S)            0.020259   
681              (B00BGA9WK2)  (B00BGA9X9W, B00ENFVJJO)            0.058590   
675              (B00BGA9WK2)  (B00BGA9X9W, B00BI83EVU)            0.058590   
693              (B00BGA9WK2)  (B00BGA9X9W, B00JK00S0S)            0.058590   
687              (B00BGA9WK2)  (B00BGA9X9W, B00HVBPRUO)            0.058590   

     consequent support   support  confidence       lift  leverage  conviction  
692            0.058590  0.000117    0.551724   9.416636  0.000105    2.100068  
674            0.058590  0.000161    0.523810   8.940199  0.000143    1.976960  
668            0.058590  0.000382    0.426230   7.274737  0.000329    1.640743  
661            0.009579  0.000103    0.400000  41.757548  0.000100    1.650701  
679            0.013763  0.000279    0.372549  27.068716  0.000269    1.571815  
..                  ...       ...         ...        ...       ...         ...  
657            0.000448  0.000110    0.005435  12.137741  0.000101    1.005014  
681            0.000844  0.000279    0.004761   5.639746  0.000229    1.003935  
675            0.000308  0.000161    0.002756   8.940199  0.000143    1.002455  
693            0.000213  0.000117    0.002005   9.416636  0.000105    1.001795  
687            0.000367  0.000103    0.001754   4.778943  0.000081    1.001389  

[702 rows x 9 columns]

In [32]:
# Add category column (to be used in viz)
apriori_rules_video_games['Category'] = 'Video Games'
apriori_rules_video_games

antecedents               consequents  antecedent support  \
692  (B00BGA9X9W, B00JK00S0S)              (B00BGA9WK2)            0.000213   
674  (B00BGA9X9W, B00BI83EVU)              (B00BGA9WK2)            0.000308   
668  (B00BGA9X9W, B00BGAA3S2)              (B00BGA9WK2)            0.000896   
661  (B003O6FV8S, B0086V5UF0)              (B0053BG26C)            0.000257   
679  (B00BGA9WK2, B00ENFVJJO)              (B00BGA9X9W)            0.000749   
..                        ...                       ...                 ...   
657              (B0050SXKU4)  (B003O6CBIG, B007XVTR5S)            0.020259   
681              (B00BGA9WK2)  (B00BGA9X9W, B00ENFVJJO)            0.058590   
675              (B00BGA9WK2)  (B00BGA9X9W, B00BI83EVU)            0.058590   
693              (B00BGA9WK2)  (B00BGA9X9W, B00JK00S0S)            0.058590   
687              (B00BGA9WK2)  (B00BGA9X9W, B00HVBPRUO)            0.058590   

     consequent support   support  confidence       lift  leverage  \
692            0.058590  0.000117    0.551724   9.416636  0.000105   
674            0.058590  0.000161    0.523810   8.940199  0.000143   
668            0.058590  0.000382    0.426230   7.274737  0.000329   
661            0.009579  0.000103    0.400000  41.757548  0.000100   
679            0.013763  0.000279    0.372549  27.068716  0.000269   
..                  ...       ...         ...        ...       ...   
657            0.000448  0.000110    0.005435  12.137741  0.000101   
681            0.000844  0.000279    0.004761   5.639746  0.000229   
675            0.000308  0.000161    0.002756   8.940199  0.000143   
693            0.000213  0.000117    0.002005   9.416636  0.000105   
687            0.000367  0.000103    0.001754   4.778943  0.000081   

     conviction     Category  
692    2.100068  Video Games  
674    1.976960  Video Games  
668    1.640743  Video Games  
661    1.650701  Video Games  
679    1.571815  Video Games  
..          ...          ...  
657    1.005014  Video Games  
681    1.003935  Video Games  
675    1.002455  Video Games  
693    1.001795  Video Games  
687    1.001389  Video Games  

[702 rows x 10 columns]

In [33]:
apriori_rules_video_games.to_csv(r'C:\Users\li_mi\Class\final_project\dev\apriori_rules_video_games.csv', index = False)

Interpretaion of 1st row of results:
    
B00JK00S0S: The Last of Us Remastered - PlayStation 4 

B00BGA9X9W: DualShock 4 Wireless Controller for PlayStation 4 - Jet Black 

B00BGA9WK2: PlayStation 4 500GB Console

### Watches Analysis 

In [34]:
#  Upload df
df_watches = pd.read_sql_query('''SELECT * FROM watches_apriori_analysis''', con=cnx)
df_watches.head()

customer_id  product_id  quantity
0     27324930  B00DKYC7TK         1
1      7211452  B000EQS1JW         1
2       912779  B005JVP0FU         1
3       805483  B000JQJS6M         1
4     32555369  B00NC8PMUK         1

In [35]:
# Create pivot table to run algorithm 
apriori_watches = df_watches.groupby(["customer_id","product_id"])["quantity"].sum().unstack().reset_index().fillna(0).set_index("customer_id")

In [36]:
# Cleaning Apriori table to remove unnecessary numbers
apriori_watches = apriori_watches.applymap(encode_units)

In [37]:
# Running apriori algorithm on cleaned dataset. 
item_association_watches = apriori(apriori_watches, min_support=0.0001, use_colnames=True)

In [38]:
#  Apriori association results table and confidence levels. 
apriori_rules_watches = association_rules(item_association_watches, metric="lift", min_threshold=1)

In [39]:
# Show output by descending order of confidence level
apriori_rules_watches  = apriori_rules_watches.sort_values(["confidence"], ascending=False)
apriori_rules_watches

antecedents               consequents  antecedent support  \
206  (B004P0UUE2, B005KKLEWS)              (B004P0UUBK)            0.000275   
194  (B003EKIS4S, B003EKNMAI)              (B008D902Q2)            0.000286   
205  (B004P0UUBK, B005KKLEWS)              (B004P0UUE2)            0.000296   
200  (B003EKIU3W, B003EKNMAI)              (B008D902Q2)            0.000337   
198  (B008D902Q2, B003EKIU3W)              (B003EKNMAI)            0.000418   
..                        ...                       ...                 ...   
201              (B008D902Q2)  (B003EKIU3W, B003EKNMAI)            0.017822   
123              (B005JVP0LE)              (B004D35W8A)            0.019740   
109              (B004YM2FV2)              (B003S7T8NM)            0.028115   
44               (B000T9VK56)              (B000HFRO8O)            0.042540   
154              (B004YM2FV2)              (B00LMP7XK2)            0.028115   

     consequent support   support  confidence        lift  leverage  \
206            0.004040  0.000122    0.444444  110.016835  0.000121   
194            0.017822  0.000122    0.428571   24.047346  0.000117   
205            0.004458  0.000122    0.413793   92.819380  0.000121   
200            0.017822  0.000122    0.363636   20.403809  0.000116   
198            0.007753  0.000122    0.292683   37.750321  0.000119   
..                  ...       ...         ...         ...       ...   
201            0.000337  0.000122    0.006869   20.403809  0.000116   
123            0.004142  0.000122    0.006202    1.497308  0.000041   
109            0.004254  0.000153    0.005443    1.279419  0.000033   
44             0.003387  0.000224    0.005276    1.557706  0.000080   
154            0.003724  0.000143    0.005080    1.364246  0.000038   

     conviction  
206    1.792728  
194    1.718812  
205    1.698277  
200    1.543423  
198    1.402832  
..          ...  
201    1.006577  
123    1.002073  
109    1.001195  
44     1.001899  
154    1.001363  

[210 rows x 9 columns]

In [40]:
# Add category column (to be used in viz)
apriori_rules_watches['Category'] = 'Watches'
apriori_rules_watches

antecedents               consequents  antecedent support  \
206  (B004P0UUE2, B005KKLEWS)              (B004P0UUBK)            0.000275   
194  (B003EKIS4S, B003EKNMAI)              (B008D902Q2)            0.000286   
205  (B004P0UUBK, B005KKLEWS)              (B004P0UUE2)            0.000296   
200  (B003EKIU3W, B003EKNMAI)              (B008D902Q2)            0.000337   
198  (B008D902Q2, B003EKIU3W)              (B003EKNMAI)            0.000418   
..                        ...                       ...                 ...   
201              (B008D902Q2)  (B003EKIU3W, B003EKNMAI)            0.017822   
123              (B005JVP0LE)              (B004D35W8A)            0.019740   
109              (B004YM2FV2)              (B003S7T8NM)            0.028115   
44               (B000T9VK56)              (B000HFRO8O)            0.042540   
154              (B004YM2FV2)              (B00LMP7XK2)            0.028115   

     consequent support   support  confidence        lift  leverage  \
206            0.004040  0.000122    0.444444  110.016835  0.000121   
194            0.017822  0.000122    0.428571   24.047346  0.000117   
205            0.004458  0.000122    0.413793   92.819380  0.000121   
200            0.017822  0.000122    0.363636   20.403809  0.000116   
198            0.007753  0.000122    0.292683   37.750321  0.000119   
..                  ...       ...         ...         ...       ...   
201            0.000337  0.000122    0.006869   20.403809  0.000116   
123            0.004142  0.000122    0.006202    1.497308  0.000041   
109            0.004254  0.000153    0.005443    1.279419  0.000033   
44             0.003387  0.000224    0.005276    1.557706  0.000080   
154            0.003724  0.000143    0.005080    1.364246  0.000038   

     conviction Category  
206    1.792728  Watches  
194    1.718812  Watches  
205    1.698277  Watches  
200    1.543423  Watches  
198    1.402832  Watches  
..          ...      ...  
201    1.006577  Watches  
123    1.002073  Watches  
109    1.001195  Watches  
44     1.001899  Watches  
154    1.001363  Watches  

[210 rows x 10 columns]

In [41]:
apriori_rules_watches.to_csv(r'C:\Users\li_mi\Class\final_project\dev\apriori_rules_watches.csv', index = False)

higher confidence analysis:

XOXO Women's XO5429 Rhinestone-Accented Two-Tone Bracelet Watch (B005KKLEWS)

XOXO Women's XO5302A Rhinestone-Accented Gold-Tone Bracelet Watch (B004P0UUE2)

XOXO Women's XO5301A Rhinestone-Accented Silver-Tone Bracelet Watch (B004P0UUBK)

In [42]:
# Concatenate dfs for visualization 
df_viz_2 = pd.concat([apriori_rules_videos,apriori_rules_pcare,apriori_rules_video_games,apriori_rules_watches], axis=0)
df_viz_2

antecedents  \
9175                (0788802194, 0788812408, 0788812807)   
11817  (0788812408, 155890641X, 0788802194, 630440113...   
11819  (155890641X, 6300274195, 0788802194, 630440113...   
11820   (155890641X, 0788812408, 0788802194, 6300274195)   
11821   (155890641X, 0788812408, 6304401132, 6300274195)   
...                                                  ...   
201                                         (B008D902Q2)   
123                                         (B005JVP0LE)   
109                                         (B004YM2FV2)   
44                                          (B000T9VK56)   
154                                         (B004YM2FV2)   

                                consequents  antecedent support  \
9175   (6304401132, 0788805533, 0788806270)            0.000165   
11817                          (6300274195)            0.000165   
11819                          (0788812408)            0.000165   
11820              (6304401132, 0788812807)            0.000165   
11821              (0788802194, 0788812807)            0.000165   
...                                     ...                 ...   
201                (B003EKIU3W, B003EKNMAI)            0.017822   
123                            (B004D35W8A)            0.019740   
109                            (B003S7T8NM)            0.028115   
44                             (B000HFRO8O)            0.042540   
154                            (B00LMP7XK2)            0.028115   

       consequent support   support  confidence         lift  leverage  \
9175             0.000165  0.000165    1.000000  6052.000000  0.000165   
11817            0.001157  0.000165    1.000000   864.571429  0.000165   
11819            0.003139  0.000165    1.000000   318.526316  0.000165   
11820            0.000165  0.000165    1.000000  6052.000000  0.000165   
11821            0.000165  0.000165    1.000000  6052.000000  0.000165   
...                   ...       ...         ...          ...       ...   
201              0.000337  0.000122    0.006869    20.403809  0.000116   
123              0.004142  0.000122    0.006202     1.497308  0.000041   
109              0.004254  0.000153    0.005443     1.279419  0.000033   
44               0.003387  0.000224    0.005276     1.557706  0.000080   
154              0.003724  0.000143    0.005080     1.364246  0.000038   

       conviction Category  
9175          inf   Videos  
11817         inf   Videos  
11819         inf   Videos  
11820         inf   Videos  
11821         inf   Videos  
...           ...      ...  
201      1.006577  Watches  
123      1.002073  Watches  
109      1.001195  Watches  
44       1.001899  Watches  
154      1.001363  Watches  

[22436 rows x 10 columns]

In [44]:
# Filter results confidence level > 60%
df_viz_2 = df_viz_2[df_viz_2['confidence'] >= 0.6]
df_viz_2

antecedents  \
9175                (0788802194, 0788812408, 0788812807)   
11817  (0788812408, 155890641X, 0788802194, 630440113...   
11819  (155890641X, 6300274195, 0788802194, 630440113...   
11820   (155890641X, 0788812408, 0788802194, 6300274195)   
11821   (155890641X, 0788812408, 6304401132, 6300274195)   
...                                                  ...   
7767                            (6304401132, 0788806270)   
3883                            (6304500831, 0788812807)   
3872                            (6304500831, 0788812807)   
644                             (155890641X, 0788802194)   
651                             (155890641X, 0788802194)   

                                consequents  antecedent support  \
9175   (6304401132, 0788805533, 0788806270)            0.000165   
11817                          (6300274195)            0.000165   
11819                          (0788812408)            0.000165   
11820              (6304401132, 0788812807)            0.000165   
11821              (0788802194, 0788812807)            0.000165   
...                                     ...                 ...   
7767   (0788812408, 0788812807, 6300274195)            0.000248   
3883               (6302794331, 6303314015)            0.000248   
3872               (155890641X, 6302158095)            0.000248   
644                            (6300274195)            0.000413   
651                            (6304401132)            0.000413   

       consequent support   support  confidence         lift  leverage  \
9175             0.000165  0.000165    1.000000  6052.000000  0.000165   
11817            0.001157  0.000165    1.000000   864.571429  0.000165   
11819            0.003139  0.000165    1.000000   318.526316  0.000165   
11820            0.000165  0.000165    1.000000  6052.000000  0.000165   
11821            0.000165  0.000165    1.000000  6052.000000  0.000165   
...                   ...       ...         ...          ...       ...   
7767             0.000165  0.000165    0.666667  4034.666667  0.000165   
3883             0.000165  0.000165    0.666667  4034.666667  0.000165   
3872             0.000248  0.000165    0.666667  2689.777778  0.000165   
644              0.001157  0.000248    0.600000   518.742857  0.000247   
651              0.001487  0.000248    0.600000   403.466667  0.000247   

       conviction Category  
9175          inf   Videos  
11817         inf   Videos  
11819         inf   Videos  
11820         inf   Videos  
11821         inf   Videos  
...           ...      ...  
7767     2.999504   Videos  
3883     2.999504   Videos  
3872     2.999256   Videos  
644      2.497108   Videos  
651      2.496282   Videos  

[17181 rows x 10 columns]

In [46]:
df_viz_2  = df_viz_2.drop(['antecedent support', 'consequent support', 'support','lift','leverage','conviction'], axis=1)
df_viz_2

antecedents  \
9175                (0788802194, 0788812408, 0788812807)   
11817  (0788812408, 155890641X, 0788802194, 630440113...   
11819  (155890641X, 6300274195, 0788802194, 630440113...   
11820   (155890641X, 0788812408, 0788802194, 6300274195)   
11821   (155890641X, 0788812408, 6304401132, 6300274195)   
...                                                  ...   
7767                            (6304401132, 0788806270)   
3883                            (6304500831, 0788812807)   
3872                            (6304500831, 0788812807)   
644                             (155890641X, 0788802194)   
651                             (155890641X, 0788802194)   

                                consequents  confidence Category  
9175   (6304401132, 0788805533, 0788806270)    1.000000   Videos  
11817                          (6300274195)    1.000000   Videos  
11819                          (0788812408)    1.000000   Videos  
11820              (6304401132, 0788812807)    1.000000   Videos  
11821              (0788802194, 0788812807)    1.000000   Videos  
...                                     ...         ...      ...  
7767   (0788812408, 0788812807, 6300274195)    0.666667   Videos  
3883               (6302794331, 6303314015)    0.666667   Videos  
3872               (155890641X, 6302158095)    0.666667   Videos  
644                            (6300274195)    0.600000   Videos  
651                            (6304401132)    0.600000   Videos  

[17181 rows x 4 columns]

In [47]:
df_viz_2.to_csv(r'C:\Users\li_mi\Class\final_project\dev\df_viz_2.csv', index = False)